crates/bitwarden-vault/src/cipher/cipher_risk_client.rs

```diff
@@ -0,0 +1,778 @@
+    /// Penalizes passwords that contain parts of the username/email.
+    fn calculate_password_strength(password: &str, username: Option<&str>) -> u8 {
+        let mut user_inputs = Vec::new();
+
+        // Extract meaningful parts from username field
+        if let Some(username) = username {
+            user_inputs.extend(Self::extract_user_inputs(username));
+        }
+
+        // Call zxcvbn with cipher-specific inputs only (no "bitwarden" globals)
+        let inputs_refs: Vec<&str> = user_inputs.iter().map(|s| s.as_str()).collect();
+        zxcvbn::zxcvbn(password, &inputs_refs).score().into()
+    }
+
+    /// Extract meaningful tokens from username/email for password penalization.
+    ///
+    /// Handles both email addresses and plain usernames:
+    /// - For emails: extracts and tokenizes the local part (before @)
+    /// - For usernames: tokenizes the entire string
+    /// - Splits on non-alphanumeric characters and converts to lowercase
+    fn extract_user_inputs(username: &str) -> Vec<String> {
+        // Check if it's email-like (contains @)
+        if let Some((local_part, _domain)) = username.split_once('@') {
+            // Email: extract local part tokens
+            local_part
+                .trim()
+                .to_lowercase()
+                .split(|c: char| !c.is_alphanumeric())
+                .filter(|s| !s.is_empty())
+                .map(str::to_owned)
+                .collect()
+        } else {
+            // Username: split on non-alphanumeric
+            username
+                .trim()
+                .to_lowercase()
+                .split(|c: char| !c.is_alphanumeric())
+                .filter(|s| !s.is_empty())
+                .map(str::to_owned)
+                .collect()
+        }
+    }
+
+    /// Hash password with SHA-1 and split into prefix/suffix for k-anonymity.
+    ///
+    /// Returns a tuple of (prefix: first 5 chars, suffix: remaining chars).
+    fn hash_password_for_hibp(password: &str) -> (String, String) {
+        use sha1::{Digest, Sha1};
+
+        let hash = Sha1::digest(password.as_bytes());
+        let hash_hex = format!("{:X}", hash);
+        let (prefix, suffix) = hash_hex.split_at(5);
+        (prefix.to_string(), suffix.to_string())
+    }
+
+    /// Parse HIBP API response to find password hash and return breach count.
+    ///
+    /// Response format: "SUFFIX:COUNT\r\n..." (e.g.,
+    /// "0018A45C4D1DEF81644B54AB7F969B88D65:3\r\n...").
+    /// Returns the number of times the password appears in breaches (0 if not found).
+    fn parse_hibp_response(response: &str, target_suffix: &str) -> u32 {
+        for line in response.lines() {
+            if let Some((hash_suffix, count_str)) = line.split_once(':') {
+                if hash_suffix.eq_ignore_ascii_case(target_suffix) {
+                    return count_str.trim().parse().unwrap_or(0);
+                }
+            }
+        }
+        0
```

Comment:
🎨 Could use iterator methods
```suggestion
        response
            .lines()
            .filter_map(|l| l.split_once(':'))
            .find(|(hash_suffix, _)| hash_suffix.eq_ignore_ascii_case(target_suffix))
            .and_then(|(_, count_str)| count_str.trim().parse().ok())
            .unwrap_or(0)
```

Reply:
[1ea9850](https://github.com/bitwarden/sdk-internal/pull/499/commits/1ea98508fd1069a84e047325bab6f67e6c8c2ef5)